### ISSUE TITLE
Once a file is opened with FileChooser, the permissions remain permanently in flatpak

### ISSUE THREAD

**Issue by gentoo-root** [flatpak/xdg-desktop-portal#1349]
### Operating System

Arch Linux

### XDG Desktop Portal version

1.18

### XDG Desktop Portal version (Other)

1.18.2

### Desktop Environment

GNOME

### Expected Behavior

The documents entries should be removed from `flatpak permission-show ...` at some point, for example, on application restart, or (better) when the application is done working with the file.

### Current Behavior

Once a file is opened with FileChooser in a flatpak application, that application gets permanent permissions to access that file. This poses a security risk (the application can access the file when the user doesn't expect it) and a privacy risk (it's possible to dump a full history of all opened file names, even when those files no longer exist; clearing this "history" isn't obvious).

### Steps to Reproduce

1. Using Telegram installed with Flatpak, send a file to someone.
2. Observe the file added to `flatpak permission-show -v org.telegram.desktop`
3. The permission is never removed from that list anymore.

### Anything else we should know?

I saw https://github.com/flatpak/xdg-desktop-portal/issues/689 and https://github.com/flatpak/flatpak/issues/4303, but those seem to refer to the FDs not being released, while I'm reporting flatpak permissions not being removed.

**Comment by Mikenux** [comment]
Hello @gentoo-root!

This isn't new behavior. This allows to have a recent files list (where files are accessible) and to restore files.

Regarding the privacy risk, an app can already keep an history of the files it accessed, no?

Regarding the security risk, logically they can access files again without informing the user of that.

The problem if we remove this is that we we'll need to ask to make them "permanently" accessible in case the user wants a functional recent file list (i.e. that opens the files) or wants session restore, which is common in apps. Here, why asking for something that is "commonly" expected? The other ways is to ask when writing files and when sharing files (network portal with relevant options), while it also depends on static permissions an app has.

**Comment by rusty-snake** [comment]
> The FileChooser UI gives an impression that I'm sharing the file with the application only once, but in fact the access remains permanent.

That's not a problem of the Permission/FileChooser-portal but of the portal implementation.

**Comment by rusty-snake** [comment]
No, improving a UI that gives you misleading impression needs to be fixed by the protal implementations.

**Comment by Mikenux** [comment]
In my opinion, informing about this in the file chooser UIs will clutter them unnecessarily.

To remove this permanent access, we then need to know how to properly ask users for it when apps need it.

If permanent access is kept, ask your DE for the ability to remove this file access permission. For example, for GNOME, this might be a feature to add in Settings > Privacy & Security.

> 1. A proprietary video game invokes `flatpak permission-show -v org.telegram.desktop` and finds out what sort of files I sent to people on Telegram.

Is this currently possible? If that's the case, isn't it up to system policy to deal with that?

> 2\. Then an adversary gets his computer for 10 seconds to type `flatpak permission-show -v org.telegram.desktop` and finds out some compromising file names. It was never obvious to the user that they had to run `flatpak permission-reset ...`.

Are file access actually kept in the permission store once deleted? If so, perhaps open a separate request for this (i.e. remove file access when the corresponding file is deleted)? (I'm not sure if this should be asked on xdg-desktop-portal or on flatpak, but issues can be moved between issue trackers)

**Comment by smcv** [comment]
> If I can run this command in the command line, a game could do the same. Games are usually not sandboxed.

If an untrusted program can execute arbitrary code with your privileges, then that's game over for any sort of security boundary, because anything you can do, it can do - including for example reconfiguring your session bus so that it will start its own modified version of xdg-desktop-portal that has different behaviour. There is no change that could be made in xdg-desktop-portal that could possibly prevent this.

Sorry, this is just not a threat model that is solvable. If you want a subset of your apps to be treated as less-trusted (potentially malicious or compromised), then the first thing that is needed is to put up a security boundary between those apps and the trusted computing base (your desktop environment).

Possible implementations of that security boundary include Flatpak (with suitably restrictive sandboxing parameters), Snap (ditto), and fast-user-switching to a separate uid that runs its own desktop session and cannot access your sensitive personal files.

> Even Lutris flatpak allows games to run `flatpak-spawn --host ...` and break the sandbox.

If that's the case, then you cannot run untrusted software from the Lutris Flatpak app and expect to be protected from it. Flatpak apps that allow `flatpak-spawn --host` are not an effective security boundary (they are part of the trusted computing base).

**Comment by Mikenux** [comment]
> This is not an problem that can be easy solved. Programs are developed with the fact, that they have permanent access to a file, in mind (the file could of course be deleted). If we change this, we break programs.

If it has to change, maybe we can have a mechanism to know which apps are using the old way (permanent access by default) vs. the new way (no permanent access by default).

> > That is, Telegram would be able to request a single-time read access, while some text editor would be able to request a session-long read-write access, etc.
> 
> That won't work.

I have open issues about that: read-only by default, ask to write. However, I need to rewrite to better organize that.

> Most programs don't call the Portal directly. They just just the FileChooser Function of their Toolkit and the Toolkit calls the Portal. So we can't really add Options to the FileChooser Portal.

Most programs don't call the portal directly - yes, but the toolkits can implement new functions if they are implemented in the portal. Just that if there are new functions, current programs shouldn't break the time developers are updating.

---

Currently, I see only one thing that can be possibly done: Autoremove file access permission when corresponding files are deleted. But doesn't it a flatpak issue?

---

To continue on about removing permanent access to files, as already said, we need to know why this permanent access exists in the first place. @smcv?

**Comment by swick** [comment]
Giving permanent access to the file was a design decision and apps and users are expecting this behavior. I don't think this is an actionable issue. If you feel like the design should change and have an idea how to get there, maybe open a discussion?

**Comment by Mikenux** [comment]
Another thing that requires persistent access to files is perhaps if the app has a search provider. However, asking for a search provider, if persistent access is something to be requested, is easier than for recent files or session restore (I don't think there are good clues to tell the difference between the latter two).

I think we also need to consider the save/write access permission: Is having persistent access to files with read-only files sufficient as a security measure? Or would security be greatly improved if persistent access to files was a permission (i.e. not present by default), despite the presence of write access permission?
